includes/classes/Indexable/Post/Post.php

```diff
@@ -404,118 +420,120 @@ private function prepare_terms( $post ) {
+		$prepared_terms = [];
+		foreach ( $terms_dictionary as $term_to_prepare ) {
+			$taxonomy_index = $term_to_prepare['taxonomy'];
 
-			if ( ! $object_terms || is_wp_error( $object_terms ) ) {
-				continue;
+			if ( ! isset( $prepared_terms[ $taxonomy_index ] ) ) {
+				$prepared_terms[ $taxonomy_index ] = [];
 			}
 
-			$terms_dic = [];
-
-			foreach ( $object_terms as $term ) {
-				if ( ! isset( $terms_dic[ $term->term_id ] ) ) {
-					$terms_dic[ $term->term_id ] = array(
-						'term_id'          => $term->term_id,
-						'slug'             => $term->slug,
-						'name'             => $term->name,
-						'parent'           => $term->parent,
-						'term_taxonomy_id' => $term->term_taxonomy_id,
-						'term_order'       => (int) $this->get_term_order( $term->term_taxonomy_id, $post->ID ),
-					);
-					if ( $allow_hierarchy ) {
-						$terms_dic = $this->get_parent_terms( $terms_dic, $term, $taxonomy->name, $post->ID );
-					}
-				}
-			}
-			$terms[ $taxonomy->name ] = array_values( $terms_dic );
+			unset( $term_to_prepare['taxonomy'] );
+			$prepared_terms[ $taxonomy_index ][] = $term_to_prepare;
 		}
 
-		return $terms;
+		return $prepared_terms;
 	}
 
 	/**
-	 * Recursively get all the ancestor terms of the given term
+	 * Recursively get all the ancestor terms of the given term.
 	 *
-	 * @param array   $terms     Terms array
-	 * @param WP_Term $term      Current term
-	 * @param string  $tax_name  Taxonomy
-	 * @param int     $object_id Post ID
+	 * @param array   $terms_dictionary Terms array
+	 * @param WP_Term $term_parent_id   Previous term's parent ID
+	 * @param string  $taxonomy         Taxonomy
 	 *
 	 * @return array
 	 */
-	private function get_parent_terms( $terms, $term, $tax_name, $object_id ) {
-		$parent_term = get_term( $term->parent, $tax_name );
-		if ( ! $parent_term || is_wp_error( $parent_term ) ) {
-			return $terms;
+	private function get_parent_terms( $terms_dictionary, $term_parent_id, $taxonomy ) {
+		$parent_term = get_term( $term_parent_id, $taxonomy );
+
+		if ( empty( $parent_term ) || is_wp_error( $parent_term ) ) {
+			return $terms_dictionary;
 		}
-		if ( ! isset( $terms[ $parent_term->term_id ] ) ) {
-			$terms[ $parent_term->term_id ] = array(
-				'term_id'    => $parent_term->term_id,
-				'slug'       => $parent_term->slug,
-				'name'       => $parent_term->name,
-				'parent'     => $parent_term->parent,
-				'term_order' => $this->get_term_order( $parent_term->term_taxonomy_id, $object_id ),
-			);
+
+		if ( ! isset( $terms_dictionary[ $parent_term->term_taxonomy_id ] ) ) {
+			$terms_dictionary[ $parent_term->term_taxonomy_id ] = [
+				'term_id'          => $parent_term->term_id,
+				'slug'             => $parent_term->slug,
+				'name'             => $parent_term->name,
+				'parent'           => $parent_term->parent,
+				'term_taxonomy_id' => $parent_term->term_taxonomy_id,
+				'taxonomy'         => $parent_term->taxonomy, // Note: This is removed before it's indexed.
```

Comment:
Let's adjust the note to include the reason for it being added in the first place.